# OpenCV:

OpenCV is a python library designed to solve computer vision problems. It is used for Facial Recognition, Dectecting Objects, Motion Detection, Vehicle Number Plate Detection, etc.

All the OpenCV array structures are converted to and from Numpy arrays. This makes it easier to integrate it with other libraries that uses Numpy.(Example: Scipy, MatplotLib, etc)

In [1]:
#pip install opencv-python

In [2]:
import cv2

In [3]:
#Color Image:
img_rgb = cv2.imread('avengers.jpg',1)

In [4]:
# Black n White (Gray Scale) image:
img_bw = cv2.imread('avengers.jpg',0)

### Shape of the Image:

In [5]:
img_rgb.shape

(750, 1000, 3)

#### 3 is for 3 RGB channels

In [6]:
img_bw.shape

(750, 1000)

#### In Black n White image there's only one channel

## Note: Python stores the image as a Numpy array / matrix of numbers. For color image 3D array and for BW 2D array.

In [7]:
# 3D matrix for color image
img_rgb

array([[[169, 169, 145],
        [176, 175, 154],
        [180, 179, 159],
        ...,
        [118,  86,  67],
        [125,  95,  76],
        [133, 103,  84]],

       [[173, 172, 151],
        [175, 174, 154],
        [175, 174, 154],
        ...,
        [121,  91,  74],
        [116,  86,  69],
        [123,  95,  78]],

       [[173, 173, 155],
        [170, 170, 152],
        [172, 170, 152],
        ...,
        [123,  94,  79],
        [109,  80,  65],
        [117,  88,  73]],

       ...,

       [[ 92,  67,  65],
        [ 87,  59,  58],
        [ 99,  68,  71],
        ...,
        [ 52,  30,  42],
        [ 51,  32,  41],
        [ 43,  26,  35]],

       [[ 99,  74,  72],
        [ 88,  60,  59],
        [ 91,  60,  63],
        ...,
        [ 52,  30,  42],
        [ 51,  31,  43],
        [ 44,  27,  38]],

       [[ 88,  63,  61],
        [ 87,  59,  58],
        [ 90,  59,  62],
        ...,
        [ 52,  30,  42],
        [ 51,  31,  43],
        [ 44,  26,  39]]

In [8]:
# 2D matrix for Black n White Image
img_bw

array([[162, 169, 173, ...,  84,  93, 101],
       [166, 168, 168, ...,  89,  84,  93],
       [168, 165, 165, ...,  93,  79,  87],
       ...,
       [ 69,  62,  72, ...,  36,  37,  31],
       [ 76,  63,  64, ...,  36,  37,  32],
       [ 65,  62,  63, ...,  36,  37,  32]], dtype=uint8)

### Display Image

In [9]:
#Color Image:
cv2.imshow("avengers_rgb", img_rgb) # avengers_rgb will be the name of the window
cv2.waitKey(0) # waitkey(0) will wait until user presses any key
cv2.destroyAllWindows() # closes the window based on waitKey parameter

In [10]:
#BW Image:
cv2.imshow("avengers_bw", img_bw)
cv2.waitKey(2000) # waitkey(2000) will wait for 2000 mSec and then will automatically destroys the window
cv2.destroyAllWindows()

### Resizing the Image:

In [11]:
resize_img = cv2.resize(img_rgb, (450,300))

In [12]:
cv2.imshow("avengers_resized_rgb", resize_img)
cv2.waitKey(0)
cv2.destroyAllWindows() 

# Face Detection: (Image)

In [13]:
#Create a Cascade Classifier object:
cascade_classifier = cv2.CascadeClassifier("haarcascades/haarcascade_frontalface_default.xml")


img = cv2.imread('avengers.jpg',1)

#Searching the co-ordinated of the image:
faces = cascade_classifier.detectMultiScale(img, scaleFactor = 1.05, minNeighbors=5)
# scaleFactor: Decreases the shape value by 5% until the face is found. Smaller the value greater the accuracy.

print(type(faces))
print(faces)

for x,y,w,h in faces:
    img = cv2.rectangle(img, (x,y),(x+w,y+h),(0,255,0),3)
    # (x,y),(x+w,y+h),(0,255,0) => RGB value of the rectangle outline and 3 is for width of the rectangle

cv2.imshow("Image Face Detection", img)

cv2.waitKey(0)

cv2.destroyAllWindows()

<class 'numpy.ndarray'>
[[821 386  37  37]
 [778 567  38  38]
 [132 546  53  53]
 [937 485  48  48]
 [650 112  77  77]
 [109 339  38  38]
 [213 485  59  59]
 [269 125  80  80]
 [471 282  83  83]
 [258 301 110 110]
 [430  28  85  85]
 [664 329  80  80]
 [281 345  74  74]]


# These are basically coordinates of the face:

<class 'numpy.ndarray'>

[[821 386  37  37]

[778 567  38  38]

[937 485  48  48]

[213 485  59  59]

[132 546  53  53]

[109 339  38  38]

[650 112  77  77]

[471 282  83  83]

[269 125  80  80]

[258 301 110 110]

[430  28  85  85]

[664 329  80  80]

[281 345  74  74]]

# Capturing Video:

In [14]:
capture = cv2.VideoCapture(0) # 0 is for builtin camera or we can also provide path of the video.
capture.release() #this will release the camera in some miliseconds.

#### when we execute above code the camera starts for split of a second and turns off to avoid this we'll be adding some delay

In [15]:
import time

In [16]:
capture = cv2.VideoCapture(0)
time.sleep(3) #this will stop the script for 3 seconds.
capture.release()

In [17]:
capture = cv2.VideoCapture(0)

check, frame = capture.read()

print(check) #it is a boolean data type, returns True if python is able to read the VideoCapture object
print(frame) #it is a Numpy array, represents the first image that video captures.

time.sleep(3)
capture.release()

True
[[[54 51 94]
  [55 48 91]
  [55 46 88]
  ...
  [47 38 35]
  [56 34 29]
  [55 32 27]]

 [[50 50 91]
  [50 48 87]
  [51 47 84]
  ...
  [50 36 41]
  [52 34 36]
  [51 32 33]]

 [[45 49 81]
  [49 50 82]
  [51 48 81]
  ...
  [53 37 42]
  [52 35 39]
  [51 34 37]]

 ...

 [[15  5 68]
  [13  7 69]
  [10  8 67]
  ...
  [20 24 47]
  [25 24 42]
  [25 22 40]]

 [[ 6  9 49]
  [ 4  9 50]
  [ 3 10 51]
  ...
  [17 23 46]
  [21 24 42]
  [22 24 42]]

 [[10 11 29]
  [10 11 29]
  [ 8 11 29]
  ...
  [26 23 42]
  [24 22 41]
  [24 22 41]]]


### Displaying video only for 3 seconds:

In [18]:
capture = cv2.VideoCapture(0)

check, frame = capture.read()

time.sleep(3)

cv2.imshow("frame",frame)

cv2.waitKey(0)

capture.release()
cv2.destroyAllWindows()

### To capture a video we need to create a loop to show multiple frames continuously

### Capturing video from Camera:

In [19]:
capture = cv2.VideoCapture(0)

a = 1

while True:
    a = a+1
    
    check, frame = capture.read()
    
    print(frame)
    
    gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    
    cv2.imshow("frame",gray)
    
    key = cv2.waitKey(1)
    
    if key == ord('q'): #'q' key will destroy the window
        break
        
print(a) #this will display the number of frames
capture.release()
cv2.destroyAllWindows()

[[[136 139 159]
  [130 139 155]
  [124 138 150]
  ...
  [138 124  98]
  [145 123 100]
  [146 123 100]]

 [[138 137 169]
  [131 132 167]
  [131 132 166]
  ...
  [132 121 111]
  [138 121 107]
  [139 121 107]]

 [[140 135 176]
  [133 131 175]
  [134 134 175]
  ...
  [135 122 112]
  [138 121 109]
  [138 120 107]]

 ...

 [[ 90  49 160]
  [ 76  54 162]
  [ 60  56 161]
  ...
  [ 72  68  64]
  [ 74  76  69]
  [ 73  76  69]]

 [[ 71  55 143]
  [ 66  61 144]
  [ 54  56 138]
  ...
  [ 36  32  27]
  [ 81  79  77]
  [ 84  83  81]]

 [[ 67  46 137]
  [ 57  37 135]
  [ 34  16 123]
  ...
  [ 76  79  72]
  [ 80  77  89]
  [ 76  73  87]]]
[[[123 146 118]
  [121 145 131]
  [132 153 151]
  ...
  [130 119 117]
  [147 117 106]
  [156 125 112]]

 [[128 145 139]
  [125 141 137]
  [129 141 138]
  ...
  [137 117 121]
  [146 109 113]
  [158 118 122]]

 [[136 143 145]
  [132 141 143]
  [131 140 141]
  ...
  [141 117 120]
  [147 110 114]
  [156 117 121]]

 ...

 [[ 43  73 134]
  [ 43  68 143]
  [ 51  70 157]
  ..

[[[123 139 156]
  [115 140 156]
  [119 148 161]
  ...
  [116 121 117]
  [112 124 112]
  [115 128 115]]

 [[124 140 157]
  [119 139 152]
  [124 145 152]
  ...
  [116 122 112]
  [112 125  99]
  [117 131 103]]

 [[121 143 137]
  [124 145 138]
  [128 145 139]
  ...
  [113 122 109]
  [111 125 103]
  [113 128 105]]

 ...

 [[ 64  51 129]
  [ 71  62 127]
  [ 70  70 122]
  ...
  [ 74  76  79]
  [ 68  71  87]
  [ 68  72  89]]

 [[ 61  45 126]
  [ 66  61 127]
  [ 60  68 120]
  ...
  [ 55  47  53]
  [ 65  68  86]
  [ 70  74  93]]

 [[ 53  47 102]
  [ 59  60 105]
  [ 59  67 103]
  ...
  [ 69  69  76]
  [ 67  73  84]
  [ 74  82  93]]]
[[[118 138 163]
  [120 139 162]
  [118 135 151]
  ...
  [127 122 108]
  [119 117  96]
  [119 117  96]]

 [[123 140 152]
  [124 139 150]
  [129 141 149]
  ...
  [128 124 101]
  [122 118  96]
  [123 119  97]]

 [[125 140 136]
  [126 140 136]
  [131 143 140]
  ...
  [120 123  97]
  [124 122  95]
  [122 118  91]]

 ...

 [[ 60  67 125]
  [ 64  63 125]
  [ 68  62 128]
  ..

[[[154 140 142]
  [152 141 145]
  [149 144 149]
  ...
  [129 127 114]
  [137 121 118]
  [139 122 120]]

 [[133 142 142]
  [141 141 142]
  [150 143 146]
  ...
  [120 124 114]
  [124 123 116]
  [125 124 118]]

 [[114 129 123]
  [140 145 138]
  [144 139 133]
  ...
  [116 125 108]
  [111 120  99]
  [109 119  97]]

 ...

 [[ 72  65 145]
  [ 71  63 146]
  [ 71  61 148]
  ...
  [ 89  73  90]
  [ 81  80 105]
  [ 75  76 102]]

 [[ 70  61 151]
  [ 69  60 151]
  [ 66  56 148]
  ...
  [ 81  69  68]
  [ 82  90  85]
  [ 73  83  77]]

 [[ 68  62 151]
  [ 67  60 153]
  [ 65  56 153]
  ...
  [ 77  73  62]
  [ 62  81  77]
  [ 60  82  79]]]
[[[127 151 147]
  [125 144 151]
  [131 144 161]
  ...
  [121 123 120]
  [132 124 111]
  [131 122 108]]

 [[130 140 135]
  [128 136 137]
  [133 138 145]
  ...
  [119 121 115]
  [124 123 107]
  [123 122 105]]

 [[133 140 128]
  [132 136 127]
  [137 138 133]
  ...
  [116 122 111]
  [119 123 107]
  [122 126 109]]

 ...

 [[ 75  57 136]
  [ 72  57 138]
  [ 70  58 139]
  ..

[[[145 138 130]
  [142 141 128]
  [138 141 127]
  ...
  [124 128 122]
  [124 128 121]
  [122 126 119]]

 [[142 143 122]
  [141 143 121]
  [140 142 124]
  ...
  [114 125 119]
  [125 124 131]
  [127 124 133]]

 [[138 143 120]
  [139 143 120]
  [140 143 122]
  ...
  [107 116 113]
  [125 116 126]
  [133 122 134]]

 ...

 [[ 75  57 165]
  [ 72  57 166]
  [ 73  59 168]
  ...
  [ 97 101 102]
  [ 88  96 100]
  [ 84  92  97]]

 [[ 70  61 153]
  [ 70  61 152]
  [ 74  64 156]
  ...
  [ 86  91  90]
  [ 72  93  93]
  [ 75  98  98]]

 [[ 78  77 113]
  [ 75  75 115]
  [ 73  73 121]
  ...
  [ 73  84  85]
  [ 72  93  93]
  [ 74  95  96]]]
[[[117 137 149]
  [125 147 157]
  [132 154 159]
  ...
  [122 126 119]
  [104 112 105]
  [113 122 114]]

 [[127 139 154]
  [123 130 149]
  [136 139 159]
  ...
  [117 126 120]
  [ 99 111 119]
  [107 120 129]]

 [[131 142 134]
  [132 140 134]
  [141 146 143]
  ...
  [116 131 124]
  [ 94 117 124]
  [ 96 121 129]]

 ...

 [[ 71  74 140]
  [ 68  73 142]
  [ 60  64 136]
  ..

[[[152 143 144]
  [151 143 145]
  [148 143 146]
  ...
  [116 116 121]
  [128 115 130]
  [128 114 130]]

 [[158 138 149]
  [155 139 147]
  [150 140 145]
  ...
  [117 120 120]
  [124 118 123]
  [122 116 121]]

 [[156 137 143]
  [150 136 142]
  [147 139 144]
  ...
  [117 121 120]
  [122 120 116]
  [122 120 116]]

 ...

 [[ 67  73 134]
  [ 52  70 138]
  [ 43  66 144]
  ...
  [110 120 126]
  [ 68  97 108]
  [ 81 112 124]]

 [[ 54  68 133]
  [ 34  62 134]
  [ 21  57 140]
  ...
  [180 189 183]
  [ 98 130 131]
  [ 76 110 112]]

 [[ 58  66 138]
  [ 61  63 138]
  [ 63  57 135]
  ...
  [226 242 246]
  [153 174 193]
  [ 88 110 130]]]
[[[147 141 148]
  [145 138 147]
  [144 140 147]
  ...
  [112 119 122]
  [113 121 118]
  [113 121 117]]

 [[140 139 125]
  [140 140 130]
  [139 143 136]
  ...
  [117 117 120]
  [119 116 121]
  [121 117 122]]

 [[138 142 124]
  [136 141 128]
  [136 143 135]
  ...
  [115 112 113]
  [121 117 117]
  [124 119 120]]

 ...

 [[ 67  70 127]
  [ 62  69 129]
  [ 60  68 133]
  ..

[[[147 141 135]
  [143 139 139]
  [142 141 145]
  ...
  [116 123 108]
  [110 123 115]
  [108 122 115]]

 [[145 142 135]
  [140 137 136]
  [140 139 144]
  ...
  [118 122 118]
  [113 115 115]
  [116 118 119]]

 [[140 138 138]
  [140 138 143]
  [142 143 149]
  ...
  [118 124 120]
  [120 125 124]
  [124 129 129]]

 ...

 [[ 67  56 105]
  [ 62  60 114]
  [ 63  67 130]
  ...
  [240 250 237]
  [237 246 229]
  [161 169 152]]

 [[ 67  62 122]
  [ 51  56 121]
  [ 37  48 120]
  ...
  [236 250 237]
  [228 246 228]
  [144 162 144]]

 [[ 51  63 130]
  [ 42  60 127]
  [ 37  58 126]
  ...
  [235 249 243]
  [201 218 216]
  [126 143 142]]]
[[[147 141 135]
  [143 139 139]
  [142 141 145]
  ...
  [116 123 108]
  [110 123 115]
  [108 122 115]]

 [[145 142 135]
  [140 137 136]
  [140 139 144]
  ...
  [118 122 118]
  [113 115 115]
  [116 118 119]]

 [[140 138 138]
  [140 138 143]
  [142 143 149]
  ...
  [118 124 120]
  [120 125 124]
  [124 129 129]]

 ...

 [[ 67  56 105]
  [ 62  60 114]
  [ 63  67 130]
  ..

[[[131 136 163]
  [132 137 154]
  [135 141 147]
  ...
  [109 118 122]
  [100 116 139]
  [107 123 149]]

 [[134 134 164]
  [130 134 159]
  [133 141 158]
  ...
  [111 117 115]
  [110 117 127]
  [110 117 130]]

 [[131 136 154]
  [126 135 152]
  [129 139 154]
  ...
  [109 114 107]
  [112 118 122]
  [112 118 123]]

 ...

 [[ 69  76 112]
  [ 64  71 116]
  [ 62  65 121]
  ...
  [242 250 236]
  [240 245 226]
  [158 162 143]]

 [[ 51  63 119]
  [ 50  61 120]
  [ 60  65 131]
  ...
  [239 249 240]
  [218 227 219]
  [141 150 142]]

 [[ 77  54 147]
  [ 83  53 148]
  [ 83  55 156]
  ...
  [231 246 252]
  [175 193 214]
  [109 128 151]]]
[[[116 138 141]
  [115 140 145]
  [131 157 166]
  ...
  [129 123 115]
  [113 123 111]
  [113 124 112]]

 [[119 136 151]
  [115 135 150]
  [131 153 169]
  ...
  [118 121 108]
  [113 122 114]
  [112 122 115]]

 [[120 135 150]
  [116 134 148]
  [125 145 159]
  ...
  [117 125 114]
  [111 122 114]
  [113 124 117]]

 ...

 [[ 74  60 159]
  [ 73  58 160]
  [ 73  58 162]
  ..

# Face Detection: (Video using WebCam)

In [20]:
capture = cv2.VideoCapture(0)

cascade_classifier = cv2.CascadeClassifier("haarcascades/haarcascade_frontalface_default.xml")

while True: #while loop to continuously read capture video
        ret, frame = capture.read()
        
        frame = cv2.cvtColor(frame,0)
        
        detections = cascade_classifier.detectMultiScale(frame)
        
        if(len(detections)>1):
            (x,y,w,h) = detections[0]
            frame = cv2.rectangle(frame,(x,y),(x+w,y+h),(255,0,0),2)
        
        cv2.imshow("frame",frame)
        if cv2.waitKey(1) & 0xFF == ord('q'):
            break
            
capture.release()
cv2.destroyAllWindows()